packages/excalidraw/components/App.tsx

```diff
@@ -6349,15 +6352,76 @@ class App extends React.Component<AppProps, AppState> {
-    this.lastPointerUpEvent = event;
+    this.removePointer(event); // always remove pointer on pointerUp
+    this.lastPointerUpEvent = event; // save last pointer up event
 
     const scenePointer = viewportCoordsToSceneCoords(
       { clientX: event.clientX, clientY: event.clientY },
       this.state,
     );
     const clicklength =
-      event.timeStamp - (this.lastPointerDownEvent?.timeStamp ?? 0);
+      event.timeStamp - (this.lastPointerDownEvent?.timeStamp ?? 0); // this is to distinguish between click and drag; a click is less than 300ms, a drag is more than 300ms; 300ms comes from the default value of the clicklength in the browser
+
+    // Ensure lastPointerDownEvent is not defined before accessing its properties
+    if (!this.lastPointerDownEvent) {
+      return; // return if lastPointerDownEvent is not defined
+    }
+
+    // Variable to track the initial pointer down position => this happens right before the pointer up event to compare with the pointer up position
+    const pointerDownPosition = {
+      x: this.lastPointerDownEvent?.clientX, // x position of the pointer down event
+      y: this.lastPointerDownEvent?.clientY, // y position of the pointer down event; clientY is a read-only property of the MouseEvent interface that returns the vertical coordinate within the application's viewport at which the event occurred (as opposed to the coordinate within the page).
+    };
+
+    // Calculate the distance moved during the inital pointer down position to the pointer up position
+    const distanceMoved = Math.sqrt(
+      Math.pow(event.clientX - pointerDownPosition.x, 2) +
+        Math.pow(event.clientY - pointerDownPosition.y, 2),
+    );
+
+    const addElement = (element: any /* : Partial<ExcalidrawElement> */) => {
+      // TODO: here we need the right ExcalidrawElement type
+      const newElement = {
+        ...element,
+        // const id = ... ?
+        // isDeleted: false ?
+        // other properties as needed
+      };
+    };
+
+    // if the distance is less than 5px and the click length is less than 300ms, then it is a click
+    const dragTreshold = 5;
+    if (distanceMoved < dragTreshold && clicklength < 300) {
+      // check the selected tool for shape creation
+      if (
+        [
+          "rectangle",
+          "ellipse",
+          "diamond",
+          "arrow",
+          "line",
+          "freedraw",
+        ].includes(this.state.activeTool.type)
+      ) {
+        const newElement = {
+          // TODO: here we need the right ExcalidrawElement type
+
+          type: this.state.activeTool.type,
+          width: 100,
+          height: 100,
+          x: scenePointer.x - 50, // Center the shape
+          y: scenePointer.y - 50, // Center the shape
+          // For later: Inlude other properties as needed (color, stroke, etc.)
+        };
+
+        // Add the new element to the scene (call the function from above)
+        addElement(newElement);
```

Comment:
Yes, this is very helpful and it makes totally sense, that there was already a function for this usecase implemented, but I just couldn´t find it 🥲 

When I write    `   this.scene.insertElement(newElement);` I get an error, because the current structure of the `newElement` does not match the expected type `ExcalidrawElement` or its subtypes (`ExcalidrawSelectionElement`, `ExcalidrawRectangleElement`, etc.) I came across the fact, that the `newElement` is missing required properties that are expected by `ExcalidrawElement`. 

**Can you help me and tell me where exactly I can find the setup for this element type in the codebase so that I can take a closer look at how it is structured?**

Reply:
I'm not sure what you're looking for without the exact error message, but here are the files related to these types:
* `packages/excalidraw/element/types.ts`
* `packages/excalidraw/element/newElement.ts`
* `packages/excalidraw/scene/Scene.ts`

Additionally you can use your IDE to "Find all references" to see where a type or function is defined and continue exploring.

That being said, **I highly recommend looking into** `createGenericElementOnPointerDown()` to see how element creation is handled in the current version.

Reply:
Hey, yeah I already looked into createGenericElementOnPointerDown(). So you can be sure, that I take all your advise very serious. I think sometimes I just have more of a problem finding the things I need or need to refer to because I´m new to this codebase and I don't know it well enough yet and I've really spent a lot of time looking. Also with the IDE. So these hints you gave me were already very helpful to find my way through.